### PR TITLE
Allow DropIndex to work on collections

### PR DESCRIPTION
**Pull request by torcolvin**
This prevents errors on DropIndexes in tests in db packages. This is not exercised until we do tests on named collections, but it works for non named collections.

## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [x] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/654/

### CODE REVIEW

File: base/collection_n1ql_common.go [hunk 1]
```diff
@@ -335,7 +335,7 @@ func getIndexMetaWithoutRetry(store N1QLStore, indexName string) (exists bool, m
 
 // DropIndex drops the specified index from the current bucket.
 func DropIndex(store N1QLStore, indexName string) error {
-	statement := fmt.Sprintf("DROP INDEX %s.`%s`", store.EscapedKeyspace(), indexName)
+	statement := fmt.Sprintf("DROP INDEX `%s` ON %s", indexName, store.EscapedKeyspace())
```
Comment (bbrks) [inline base/collection_n1ql_common.go]: The `ON` syntax is only available in Couchbase Server 7.0+ - so we might need special handling only when using collections for this (where we know we'd not be on 6.x). Or find a way to make the old syntax work with bucket and collection keyspaces.

https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/dropindex.html#keyspace-ref

**Review by bbrks — CHANGES_REQUESTED** [review]

File: base/collection_n1ql_common.go (hunk 1 above)
Comment (torcolvin) [reply base/collection_n1ql_common.go]: I changed this to the docs and spec for 6.0 - the reason this didn't get hit by tests earlier is that typically there were no  indexes in the way that the code was created, so I added a test at the end to Drop them after we created them explicitly in the test.

File: db/indexes_test.go [hunk 2]
```diff
@@ -69,6 +71,10 @@ func TestInitializeIndexes(t *testing.T) {
 
 			validateErr := validateAllIndexesOnline(b)
 			require.NoError(t, validateErr, "Error validating indexes online")
+
+			// Drop again
+			dropErr = base.DropAllIndexes(base.TestCtx(t), n1qlStore)
+			require.NoError(t, dropErr, "Error dropping all indexes")
```
Comment (bbrks) [inline db/indexes_test.go]: I think we should only do this if `test.collections == true`

The bucket indexes are expected to be kept around when pooled for the next test to use - they're not recreated in the bucket reader - only emptied

**Review by bbrks — CHANGES_REQUESTED** [review]
This change has broken `TestInitializeIndexes` because the syntax is apparently not valid for just a bucket - According to the N1QL docs it _should_ work - but getting clarification from the query team to see whether it's a server bug might be useful.

- `gsi ^TestInitializeIndexes$` https://jenkins.sgwdev.com/job/SyncGateway-Integration/650/

```
--- FAIL: TestInitializeIndexes (39.45s)
    --- FAIL: TestInitializeIndexes/xattrs=true_collections=false (1.23s)
    --- FAIL: TestInitializeIndexes/xattrs=false_collections=false (1.43s)
    --- PASS: TestInitializeIndexes/xattrs=true_collections=true (15.89s)
    --- PASS: TestInitializeIndexes/xattrs=false_collections=true (20.90s)
```

```
=== RUN   TestInitializeIndexes/xattrs=true_collections=false
2022-08-18T11:30:05.156+01:00 [INF] Created background task: "CleanAgedItems" with interval 1m0s
2022-08-18T11:30:05.156+01:00 [INF] Created background task: "InsertPendingEntries" with interval 2.5s
2022-08-18T11:30:05.157+01:00 [INF] Created background task: "CleanSkippedSequenceQueue" with interval 30m0s
2022-08-18T11:30:06.295+01:00 [INF] c:db:db Using metadata purge interval of 3.00 days for tombstone compaction.
2022-08-18T11:30:06.373+01:00 [ERR] t:TestInitializeIndexes/xattrs=true_collections=false ...failed to drop index sg_tombstones_x1 on bucket sg_int_1_1660818470158881000: parsing failure | {"statement":"DROP INDEX default:`sg_int_1_1660818470158881000`.`sg_tombstones_x1`","client_context_id":"94118061-00a7-4d4a-b548-65947cd5356c","errors":[{"code":3000,"message":"syntax error - at end of input"}],"http_status_code":400} -- base.DropAllIndexes.func1() at util_testing.go:306
2022-08-18T11:30:06.374+01:00 [ERR] t:TestInitializeIndexes/xattrs=true_collections=false ...failed to drop index sg_allDocs_x1 on bucket sg_int_1_1660818470158881000: parsing failure | {"statement":"DROP INDEX default:`sg_int_1_1660818470158881000`.`sg_allDocs_x1`","client_context_id":"73ced936-35f1-46ae-b525-9a3646c6a07f","errors":[{"code":3000,"message":"syntax error - at end of input"}],"http_status_code":400} -- base.DropAllIndexes.func1() at util_testing.go:306
2022-08-18T11:30:06.374+01:00 [ERR] t:TestInitializeIndexes/xattrs=true_collections=false ...failed to drop index sg_syncDocs_x1 on bucket sg_int_1_1660818470158881000: parsing failure | {"statement":"DROP INDEX default:`sg_int_1_1660818470158881000`.`sg_syncDocs_x1`","client_context_id":"e98ad93f-7e89-480c-ab64-36c69fce94a7","errors":[{"code":3000,"message":"syntax error - at end of input"}],"http_status_code":400} -- base.DropAllIndexes.func1() at util_testing.go:306
2022-08-18T11:30:06.375+01:00 [ERR] t:TestInitializeIndexes/xattrs=true_collections=false ...failed to drop index sg_roleAccess_x1 on bucket sg_int_1_1660818470158881000: parsing failure | {"statement":"DROP INDEX default:`sg_int_1_1660818470158881000`.`sg_roleAccess_x1`","client_context_id":"735c209a-8b94-4ce3-a936-6cdd5a2a534a","errors":[{"code":3000,"message":"syntax error - at end of input"}],"http_status_code":400} -- base.DropAllIndexes.func1() at util_testing.go:306
2022-08-18T11:30:06.375+01:00 [ERR] t:TestInitializeIndexes/xattrs=true_collections=false ...failed to drop index sg_access_x1 on bucket sg_int_1_1660818470158881000: parsing failure | {"statement":"DROP INDEX default:`sg_int_1_1660818470158881000`.`sg_access_x1`","client_context_id":"fb6f57f2-dc28-4770-812c-a2e043fa6751","errors":[{"code":3000,"message":"syntax error - at end of input"}],"http_status_code":400} -- base.DropAllIndexes.func1() at util_testing.go:306
2022-08-18T11:30:06.376+01:00 [ERR] t:TestInitializeIndexes/xattrs=true_collections=false ...failed to drop index sg_channels_x1 on bucket sg_int_1_1660818470158881000: parsing failure | {"statement":"DROP INDEX default:`sg_int_1_1660818470158881000`.`sg_channels_x1`","client_context_id":"cfe7dbfe-84a3-4e16-bb0a-ef59eefa3253","errors":[{"code":3000,"message":"syntax error - at end of input"}],"http_status_code":400} -- base.DropAllIndexes.func1() at util_testing.go:306
2022-08-18T11:30:06.386+01:00 [ERR] t:TestInitializeIndexes/xattrs=true_collections=false ...failed to drop index #primary on bucket sg_int_1_1660818470158881000: parsing failure | {"statement":"DROP INDEX default:`sg_int_1_1660818470158881000`.`#primary`","client_context_id":"456363d6-2f2d-443a-b851-a1e062a0579a","errors":[{"code":3000,"message":"syntax error - at end of input"}],"http_status_code":400} -- base.DropAllIndexes.func1() at util_testing.go:306
    indexes_test.go:63:
        	Error Trace:	indexes_test.go:63
        	Error:      	Received unexpected error:
        	            	parsing failure | {"statement":"DROP INDEX default:`sg_int_1_1660818470158881000`.`sg_tombstones_x1`","client_context_id":"94118061-00a7-4d4a-b548-65947cd5356c","errors":[{"code":3000,"message":"syntax error - at end of input"}],"http_status_code":400}
        	Test:       	TestInitializeIndexes/xattrs=true_collections=false
        	Messages:   	Error dropping all indexes
```

File: db/indexes_test.go (hunk 2 above)
Comment (torcolvin) [reply db/indexes_test.go]: The reasons that this doesn't work is that for server 6.x a keyspace is only a bucket, and for 7.x the keyspace needs to be fully qualified if you are not using the `ON` keyword.

We could create a function that gets a fully qualified keyspace:

if 6.x -> `default:bucket`
if 7.x -> `default:bucket._default._default` for unnamed collection
if 7.x -> `default:bucket.scope.collection` for named collection

Comment (torcolvin) [reply db/indexes_test.go]: As per https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/dropindex.html#keyspace-full-index, changing our `EscapedKeyspace` function to match this which will always be correct (if verbose). This seems to pass tests against 6.x (out of jenkins) and 7.x in jenkins.

**Review by bbrks — APPROVED** [review]

**Pull request merged**